fastlane/lib/fastlane/plugins/plugin_manager.rb

```diff
@@ -0,0 +1,338 @@
+module Fastlane
+  class PluginManager
+    require "bundler"
+
+    PLUGINFILE_NAME = "Pluginfile".freeze
+    DEFAULT_GEMFILE_PATH = "Gemfile".freeze
+    GEMFILE_SOURCE_LINE = "source \"https://rubygems.org\"\n"
+    FASTLANE_PLUGIN_PREFIX = "fastlane-plugin-"
+    TROUBLESHOOTING_URL = "https://github.com/fastlane/fastlane/blob/master/fastlane/docs/PluginsTroubleshooting.md"
+
+    #####################################################
+    # @!group Reading the files and their paths
+    #####################################################
+
+    def gemfile_path
+      # This is pretty important, since we don't know what kind of
+      # Gemfile the user has (e.g. Gemfile, gems.rb, or custom env variable)
+      Bundler::SharedHelpers.default_gemfile.to_s
+    rescue Bundler::GemfileNotFound
+      nil
+    end
+
+    def pluginfile_path
+      if FastlaneFolder.path
+        return File.join(FastlaneFolder.path, PLUGINFILE_NAME)
+      else
+        return nil
+      end
+    end
+
+    def gemfile_content
+      File.read(gemfile_path) if gemfile_path && File.exist?(gemfile_path)
+    end
+
+    def pluginfile_content
+      File.read(pluginfile_path) if pluginfile_path && File.exist?(pluginfile_path)
+    end
+
+    #####################################################
+    # @!group Helpers
+    #####################################################
+
+    def self.plugin_prefix
+      FASTLANE_PLUGIN_PREFIX
+    end
+
+    # Returns an array of gems that are added to the Gemfile or Pluginfile
+    def available_gems
+      return [] unless gemfile_path
+      dsl = Bundler::Dsl.evaluate(gemfile_path, nil, true)
+      return dsl.dependencies.map(&:name)
+    end
+
+    # Returns an array of fastlane plugins that are added to the Gemfile or Pluginfile
+    # The returned array contains the string with their prefixes (e.g. fastlane-plugin-xcversion)
+    def available_plugins
+      available_gems.keep_if do |current|
+        current.start_with?(self.class.plugin_prefix)
+      end
+    end
+
+    # Check if a plugin is added as dependency to either the
+    # Gemfile or the Pluginfile
+    def plugin_is_added_as_dependency?(plugin_name)
+      UI.user_error!("fastlane plugins must start with '#{self.class.plugin_prefix}' string") unless plugin_name.start_with?(self.class.plugin_prefix)
+      return available_plugins.include?(plugin_name)
+    end
+
+    #####################################################
+    # @!group Modifying dependencies
+    #####################################################
+
+    def add_dependency(plugin_name)
+      UI.user_error!("fastlane is not setup for this project, make sure you have a fastlane folder") unless pluginfile_path
+      plugin_name = self.class.plugin_prefix + plugin_name unless plugin_name.start_with?(self.class.plugin_prefix)
+
+      unless plugin_is_added_as_dependency?(plugin_name)
+        content = pluginfile_content || "# Autogenerated by fastlane\n\n"
```

Comment:
Consider adding a message like "Ensure this file is checked in to source control!"